src/coordination/coordinator_instance.cpp

```diff
@@ -362,35 +345,231 @@ auto CoordinatorInstance::UnregisterReplicationInstance(std::string instance_nam
+        inst_to_remove.StopFrequentCheck();
+
+        auto const is_main = [this](ReplicationInstance const &instance) {
+          return raft_state_.IsMain(instance.InstanceName());
+        };
+
+        if (auto curr_main = std::ranges::find_if(repl_instances_, is_main); curr_main != repl_instances_.end()) {
+          if (!curr_main->SendUnregisterReplicaRpc(instance_name)) {
+            // TODO: (andi) Restore state in the RAFT log if needed.
+            // inst_to_remove.StartFrequentCheck();
+          }
+        }
+        std::erase_if(repl_instances_, name_matches);
+      } else {
+        auto lock = std::unique_lock{coord_instance_lock_};
+        std::erase_if(repl_instances_, name_matches);
+      }
+
+      spdlog::info("Instance {} unregistered", instance_name);
+      break;
+    }
+    case RaftLogAction::SET_INSTANCE_AS_MAIN: {
+      auto const instance_name = std::get<std::string>(log_entry);
+      auto &new_main = FindReplicationInstance(instance_name);
+      auto const new_main_uuid = utils::UUID{};
+
+      if (raft_state_.IsLeader()) {
+        // Only leader was pinging ReplicationInstance so only leader needs to pause.
+        new_main.PauseFrequentCheck();
+        utils::OnScopeExit scope_exit{[&new_main] { new_main.ResumeFrequentCheck(); }};
+
+        auto const is_not_new_main = [&instance_name](ReplicationInstance const &instance) {
+          return instance.InstanceName() != instance_name;
+        };
+
+        auto alive_instances = repl_instances_ | ranges::views::filter(&ReplicationInstance::IsAlive) |
+                               ranges::views::filter(is_not_new_main);
+
+        auto const failed_to_swap = [&new_main_uuid](ReplicationInstance &instance) {
+          return !instance.SendSwapAndUpdateUUID(new_main_uuid);
+        };
+
+        if (std::ranges::any_of(alive_instances, failed_to_swap)) {
+          spdlog::error("Failed to swap uuid for instance {}.", new_main.InstanceName());
+          // TODO: (andi) What to do on network failure when appended to raft log?
+        }
+
+        auto repl_clients_info = repl_instances_ | ranges::views::filter(is_not_new_main) |
+                                 ranges::views::transform(&ReplicationInstance::ReplicationClientInfo) |
+                                 ranges::to<ReplicationClientsInfo>();
 
-// TODO: (andi) Add to the RAFT log.
-auto CoordinatorInstance::SetMainUUID(utils::UUID new_uuid) -> void { main_uuid_ = new_uuid; }
+        if (!new_main.PromoteToMainAsLeader(new_main_uuid, std::move(repl_clients_info),
+                                            &CoordinatorInstance::MainSuccessCallback,
+                                            &CoordinatorInstance::MainFailCallback)) {
+          spdlog::error("Failed to promote instance {} to main", instance_name);
+          // TODO: (andi) What to do on network failure when appended to raft log?
+        }
+        // Must be executed under the lock.
```

Comment:
I would like to sync with you why you chose this approach. I think this can fail in following way:
- the main coordinator tries failover
- sends a commit to all other coordinators
- The coordinator who tried failover doesn't succeed because of some failure on one of the replica instances
- the main coordinator dies
- new coordinator becomes main, but has wrong state - thinks we succeded in failover but we didn't